src/rdkafka_telemetry_decode.c

```diff
@@ -0,0 +1,543 @@
+        unit_test_data.current_field++;
+}
+
+static void unit_test_telemetry_error(void *opaque, const char *error, ...) {
+        char buffer[1024];
+        va_list ap;
+        va_start(ap, error);
+        rd_vsnprintf(buffer, sizeof(buffer), error, ap);
+        va_end(ap);
+        RD_UT_SAY("%s", buffer);
+}
+
+bool unit_test_telemetry(rd_kafka_telemetry_producer_metric_name_t metric_name,
+                         const char *expected_name,
+                         const char *expected_description,
+                         rd_kafka_telemetry_metric_type_t expected_type,
+                         rd_bool_t is_double) {
+        rd_kafka_t *rk = rd_calloc(1, sizeof(*rk));
+        rwlock_init(&rk->rk_lock);
+        rk->rk_type                          = RD_KAFKA_PRODUCER;
+        rk->rk_telemetry.matched_metrics_cnt = 1;
+        rk->rk_telemetry.matched_metrics =
+            rd_malloc(sizeof(rd_kafka_telemetry_producer_metric_name_t) *
+                      rk->rk_telemetry.matched_metrics_cnt);
+        rk->rk_telemetry.matched_metrics[0] = metric_name;
+        rd_strlcpy(rk->rk_name, "unittest", sizeof(rk->rk_name));
+        clear_unit_test_data();
+
+        rd_kafka_telemetry_decode_interface_t decode_interface = {
+            .decoded_string = unit_test_telemetry_decoded_string,
+            .decoded_NumberDataPoint =
+                unit_test_telemetry_decoded_NumberDataPoint,
+            .decoded_int64 = NULL,
+            .decoded_type  = unit_test_telemetry_decoded_type,
+            .decode_error  = unit_test_telemetry_error,
+            .opaque        = &unit_test_data,
+        };
+
+        TAILQ_INIT(&rk->rk_brokers);
+
+        rd_kafka_broker_t *rkb  = rd_calloc(1, sizeof(*rkb));
+        rkb->rkb_c.connects.val = 1;
+        TAILQ_INSERT_HEAD(&rk->rk_brokers, rkb, rkb_link);
+        rd_buf_t *rbuf              = rd_kafka_telemetry_encode_metrics(rk);
+        void *metrics_payload       = rbuf->rbuf_wpos->seg_p;
+        size_t metrics_payload_size = rbuf->rbuf_wpos->seg_of;
+        RD_UT_SAY("metrics_payload_size: %zu", metrics_payload_size);
+
+        RD_UT_ASSERT(metrics_payload_size != 0, "Metrics payload zero");
+
+        bool decode_status = rd_kafka_telemetry_decode_metrics(
+            &decode_interface, metrics_payload, metrics_payload_size);
+
+        RD_UT_ASSERT(decode_status == 1, "Decoding failed");
+        RD_UT_ASSERT(unit_test_data.type == expected_type,
+                     "Metric type mismatch");
+        RD_UT_ASSERT(strcmp(unit_test_data.metric_name, expected_name) == 0,
+                     "Metric name mismatch");
+        RD_UT_ASSERT(strcmp(unit_test_data.metric_description,
+                            expected_description) == 0,
+                     "Metric description mismatch");
+        if (is_double)
+                RD_UT_ASSERT(rd_dbl_eq(unit_test_data.metric_value_double, 0.0),
+                             "Metric value mismatch");
```

Comment:
Allow a higher error and with previous change it can expect a value of 1.0
```suggestion
                RD_UT_ASSERT(
                    rd_dbl_eq0(unit_test_data.metric_value_double, 1.0, 0.01),
                    "Metric value mismatch");
```

Reply:
Updated